// node-simctl/build/lib/helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SIM_RUNTIME_NAME = exports.DEFAULT_EXEC_TIMEOUT = void 0;
exports.normalizeVersion = normalizeVersion;
exports.getXcrunBinary = getXcrunBinary;
exports.uuidV4 = uuidV4;
exports.convertPlistToJson = convertPlistToJson;
const semver = __importStar(require("semver"));
const node_child_process_1 = require("node:child_process");
const node_stream_1 = require("node:stream");
exports.DEFAULT_EXEC_TIMEOUT = 10 * 60 * 1000; // ms
exports.SIM_RUNTIME_NAME = 'com.apple.CoreSimulator.SimRuntime.';
/**
 * "Normalize" the version, since iOS uses 'major.minor' but the runtimes can
 * be 'major.minor.patch'
 *
 * @param version - the string version
 * @return The version in 'major.minor' form
 * @throws {Error} If the version not parseable by the `semver` package
 */
function normalizeVersion(version) {
    const semverVersion = semver.coerce(version);
    if (!semverVersion) {
        throw new Error(`Unable to parse version '${version}'`);
    }
    return `${semverVersion.major}.${semverVersion.minor}`;
}
/**
 * @returns The xcrun binary name
 */
function getXcrunBinary() {
    return process.env.XCRUN_BINARY || 'xcrun';
}
/**
 * Generate a UUID v4
 *
 * @returns Promise resolving to UUID string
 */
async function uuidV4() {
    const uuidLib = await import('uuid');
    return uuidLib.v4();
}
/**
 * Convert plist-style output to JSON using plutil
 *
 * @param plistInput - The plist-style string to convert
 * @return Promise resolving to parsed JSON object
 * @throws {Error} If plutil fails to convert the input
 */
async function convertPlistToJson(plistInput) {
    const plutilProcess = (0, node_child_process_1.spawn)('plutil', ['-convert', 'json', '-o', '-', '-']);
    let jsonOutput = '';
    plutilProcess.stdout.on('data', (chunk) => {
        jsonOutput += chunk.toString();
    });
    const inputStream = node_stream_1.Readable.from([plistInput]);
    inputStream.pipe(plutilProcess.stdin);
    try {
        await new Promise((resolve, reject) => {
            inputStream.once('error', reject);
            plutilProcess.once('exit', (code, signal) => {
                inputStream.unpipe(plutilProcess.stdin);
                if (code === 0) {
                    resolve();
                }
                else {
                    reject(new Error(`plutil exited with code ${code}, signal ${signal}`));
                }
            });
            plutilProcess.once('error', (e) => {
                inputStream.unpipe(plutilProcess.stdin);
                reject(e);
            });
        });
    }
    catch (err) {
        plutilProcess.kill(9);
        throw new Error(`Failed to convert plist to JSON: ${err instanceof Error ? err.message : String(err)}`);
    }
    finally {
        plutilProcess.removeAllListeners();
        inputStream.removeAllListeners();
    }
    return JSON.parse(jsonOutput);
}
